Compact record constructors not working properly

When I in the OpenRewrite Git repo add a test with a [compact record constructor](https://docs.oracle.com/en/java/javase/14/language/records.html#GUID-6699E26F-4A9B-4393-A08B-1E47D4B2D263__GUID-3677B517-B179-4891-BA52-45CB774E7250) like the following:

```java
public class RecordTest implements RewriteTest {

    @Test
    void javaRecord() {
        rewriteRun(
          java(
            """
                  public record JavaRecord(String name, @Deprecated int age) {
                    public JavaRecord {
                        java.util.Objects.requireNonNull(name);
                    }
                  }
              """
          )
        );
    }

}
```

The test fails, because it expects:
```java
public record JavaRecord(String name, @Deprecated int age) {
  public JavaRecord {
      java.util.Objects.requireNonNull(Stringname,int age) {
  public JavaRecord {
      java.util.Objects.requireNonNull(name);
  }
}
```

This looks rather strange. I didn't analyze it any further.

This test fails with another error, but I suspect it has the same root cause:
```java
    @Test
    void javaRecord_2() {
        rewriteRun(
          java(
            """
                  public record Expression(String nested) {
                    public Expression {
                        java.util.Objects.requireNonNull(nested);
                    }
                  }
              """
          )
        );
    }
```
The exception being logged is:
```
Caused by: java.lang.StringIndexOutOfBoundsException: begin 120, end 117, length 117
	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:4601)
	at java.base/java.lang.String.substring(String.java:2704)
	at java.base/java.lang.String.substring(String.java:2677)
	at org.openrewrite.java.isolated.ReloadableJava17ParserVisitor.visitCompilationUnit(ReloadableJava17ParserVisitor.java:565)
	at org.openrewrite.java.isolated.ReloadableJava17ParserVisitor.visitCompilationUnit(ReloadableJava17ParserVisitor.java:71)
	at com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:614)
	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
	at org.openrewrite.java.isolated.ReloadableJava17Parser.lambda$parseInputs$0(ReloadableJava17Parser.java:178)
	... 116 more
```

--- a/rewrite-java-17/src/main/java/org/openrewrite/java/isolated/ReloadableJava17ParserVisitor.java
+++ b/rewrite-java-17/src/main/java/org/openrewrite/java/isolated/ReloadableJava17ParserVisitor.java
@@ -34,6 +34,7 @@ import org.openrewrite.internal.ListUtils;
 import org.openrewrite.internal.lang.Nullable;
 import org.openrewrite.java.JavaParsingException;
 import org.openrewrite.java.internal.JavaTypeCache;
+import org.openrewrite.java.marker.CompactConstructor;
 import org.openrewrite.java.marker.OmitParentheses;
 import org.openrewrite.java.tree.*;
 import org.openrewrite.marker.Markers;
@@ -995,12 +996,16 @@ public class ReloadableJava17ParserVisitor extends TreePathScanner<J, Space> {
                     node.getName().toString(), null, null), returnType == null ? returnTypeAnnotations : Collections.emptyList());
         }
 
-        Space paramFmt = sourceBefore("(");
-        JContainer<Statement> params = !node.getParameters().isEmpty() ?
-                JContainer.build(paramFmt, convertAll(node.getParameters(), commaDelim, t -> sourceBefore(")")),
-                        Markers.EMPTY) :
-                JContainer.build(paramFmt, singletonList(padRight(new J.Empty(randomId(), sourceBefore(")"),
-                        Markers.EMPTY), EMPTY)), Markers.EMPTY);
+        boolean isCompactConstructor = nodeSym != null && (nodeSym.flags() & Flags.COMPACT_RECORD_CONSTRUCTOR) != 0;
+        JContainer<Statement> params = JContainer.empty();
+        if (!isCompactConstructor) {
+            Space paramFmt = sourceBefore("(");
+            params = !node.getParameters().isEmpty() ?
+                    JContainer.build(paramFmt, convertAll(node.getParameters(), commaDelim, t -> sourceBefore(")")),
+                            Markers.EMPTY) :
+                    JContainer.build(paramFmt, singletonList(padRight(new J.Empty(randomId(), sourceBefore(")"),
+                            Markers.EMPTY), EMPTY)), Markers.EMPTY);
+        }
 
         JContainer<NameTree> throwss = node.getThrows().isEmpty() ? null :
                 JContainer.build(sourceBefore("throws"), convertAll(node.getThrows(), commaDelim, noDelim),
@@ -1011,11 +1016,13 @@ public class ReloadableJava17ParserVisitor extends TreePathScanner<J, Space> {
         JLeftPadded<Expression> defaultValue = node.getDefaultValue() == null ? null :
                 padLeft(sourceBefore("default"), convert(node.getDefaultValue()));
 
-        return new J.MethodDeclaration(randomId(), fmt, Markers.EMPTY,
+        J.MethodDeclaration md = new J.MethodDeclaration(randomId(), fmt, Markers.EMPTY,
                 modifierResults.getLeadingAnnotations(),
                 modifierResults.getModifiers(), typeParams,
                 returnType, name, params, throwss, body, defaultValue,
                 typeMapping.methodDeclarationType(jcMethod.sym, null));
+        md = isCompactConstructor ? md.withMarkers(md.getMarkers().addIfAbsent(new CompactConstructor(randomId()))) : md;
+        return md;
     }
 
     @Override

--- a/rewrite-java-17/src/test/java/org/openrewrite/java/tree/RecordTest.java
+++ b/rewrite-java-17/src/test/java/org/openrewrite/java/tree/RecordTest.java
@@ -36,7 +36,6 @@ public class RecordTest implements RewriteTest {
         );
     }
 
-    @Disabled("Requires parsing support for record compact constructors.")
     @Issue("https://github.com/openrewrite/rewrite/issues/2455")
     @Test
     void compactConstructor() {

--- a/rewrite-java/src/main/java/org/openrewrite/java/JavaPrinter.java
+++ b/rewrite-java/src/main/java/org/openrewrite/java/JavaPrinter.java
@@ -18,6 +18,7 @@ package org.openrewrite.java;
 import org.openrewrite.Cursor;
 import org.openrewrite.PrintOutputCapture;
 import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.java.marker.CompactConstructor;
 import org.openrewrite.java.marker.OmitParentheses;
 import org.openrewrite.java.tree.*;
 import org.openrewrite.java.tree.J.*;
@@ -721,7 +722,9 @@ public class JavaPrinter<P> extends JavaVisitor<PrintOutputCapture<P>> {
         visit(method.getReturnTypeExpression(), p);
         visit(method.getAnnotations().getName().getAnnotations(), p);
         visit(method.getName(), p);
-        visitContainer("(", method.getPadding().getParameters(), JContainer.Location.METHOD_DECLARATION_PARAMETERS, ",", ")", p);
+        if (!method.getMarkers().findFirst(CompactConstructor.class).isPresent()) {
+            visitContainer("(", method.getPadding().getParameters(), JContainer.Location.METHOD_DECLARATION_PARAMETERS, ",", ")", p);
+        }
         visitContainer("throws", method.getPadding().getThrows(), JContainer.Location.THROWS, ",", null, p);
         visit(method.getBody(), p);
         visitLeftPadded("default", method.getPadding().getDefaultValue(), JLeftPadded.Location.METHOD_DECLARATION_DEFAULT_VALUE, p);

--- a/rewrite-java/src/main/java/org/openrewrite/java/marker/CompactConstructor.java
+++ b/rewrite-java/src/main/java/org/openrewrite/java/marker/CompactConstructor.java
@@ -1,0 +1,32 @@
+/*
+ * Copyright 2022 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.marker;
+
+import lombok.Value;
+import lombok.With;
+import org.openrewrite.marker.Marker;
+
+import java.util.UUID;
+
+/**
+ * Java 14 introduced compact constructors, which does not include the `()` for parameters.
+ * This marker prevents the {@link org.openrewrite.java.JavaPrinter} from printer the parameters.
+ */
+@Value
+@With
+public class CompactConstructor implements Marker {
+    UUID id;
+}

